### PR TITLE
[operands]: Adding DaemonSet operand to deploy the swap agent

### PR DESCRIPTION
**Pull request by enp0s3**
Signed-off-by: Igor Bezukh [ibezukh@redhat.com](mailto:ibezukh@redhat.com)

What this PR does / why we need it:

The what

[wasp-agent ](https://github.com/fabiand/wasp-operator-prototype)allows burstable QoS pods to
use swap, thus allowing for higher VM workload density on the
worker nodes. The Agent is a script that delivered as DaemonSet.
The delivery is gated and its done directly by HCO operator.
The DS need privilege access to the node thus using the kubevirt-handler
service account.

The why

Swap support in kuberentes is currently beta1. In order to be enabled
natively in Openshift the swap needs to graduate to beta2 phase. The current
approach with wasp-agent aims to mimic the beta2 functionaliy.

Upgrade path

As soon as swap becomes beta2, the wasp-agent will be deprecated.
Users who would like to opt-in for higher density with swap will have
to enable the HigherDensityWithSwap FG in HCO. As soon as swap will
be natively supported in OCP, the cluster admin would be instructed
to validate proper kubelet configuration to allow transpart upgrade.

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:

```jira-ticket

```

**Release note**:

```release-note
adding DaemonSet operand to deploy the swap agent
```

### CODE REVIEW

File: controllers/operands/wasp.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+}
+
+type waspHooks struct {
+	cache *appsv1.DaemonSet
+}
+
+func (h *waspHooks) getFullCr(hc *hcov1beta1.HyperConverged) (client.Object, error) {
+	if h.cache == nil {
+		h.cache = NewWasp(hc)
+	}
```
Comment (nunnatsa) [inline controllers/operands/wasp.go:59]: If you're using cache, please also implement the `reset()` method; see here for example: https://github.com/kubevirt/hyperconverged-cluster-operator/blob/5eb460e178f7c349f483e3586ff3694ec69cdf03/controllers/operands/cdi.go#L59-L61

**Review by nunnatsa — COMMENTED** [review]
Added one last comment

File: controllers/operands/wasp.go (hunk 1 above)
Comment (enp0s3) [reply controllers/operands/wasp.go:59]: @nunnatsa Thanks! done

File: pkg/util/cluster.go [hunk 2]
```diff
@@ -33,6 +33,7 @@ type ClusterInfo interface {
 	IsControlPlaneHighlyAvailable() bool
 	IsInfrastructureHighlyAvailable() bool
 	IsConsolePluginImageProvided() bool
+	IsWaspImageProvided() bool
```
Comment (nunnatsa) [inline pkg/util/cluster.go:36]: This should not be part of this interface, IMHO (`IsConsolePluginImageProvided` should not be here as well). The interface is about the cluster HCO is running in.

The question if an image is provided or not, does not belong here.

File: tests/func-tests/wasp_test.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	tests "github.com/kubevirt/hyperconverged-cluster-operator/tests/func-tests"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/tests/flags"
+)
+
+const (
+	setWaspFGPatchTemplate = `[{"op": "replace", "path": "/spec/featureGates/enableHigherDensityWithSwap", "value": %t}]`
+)
+
+var _ = Describe("wasp-agent", Label("wasp"), Serial, Ordered, func() {
```
Comment (nunnatsa) [inline tests/func-tests/wasp_test.go:22]: Please add the new "OpenShift" label. See #2860 
```suggestion
var _ = Describe("wasp-agent", Label("wasp", "OpenShift"), Serial, Ordered, func() {
```

**Review by nunnatsa — COMMENTED** [review]
Since we need to fix the CSV, please also see two new comments inline.

File: tests/func-tests/wasp_test.go (hunk 3 above)
Comment (enp0s3) [reply tests/func-tests/wasp_test.go]: @nunnatsa Done

File: pkg/util/cluster.go (hunk 2 above)
Comment (enp0s3) [reply pkg/util/cluster.go:36]: @nunnatsa Following our conversation, we decided to defer it to follow up PR

**Pull request merged**